File: admin/backend/models/product.py
from bson import ObjectId
from datetime import datetime
from ..db import get_db

db = get_db()
products_collection = db["products"]


class Product:

    def __init__(
        self,
        id=None,
        sku=None,
        name=None,
        description="",
        category="General",
        root=None,
        child=None,
        unit="Unidad",

        price_sale=0.0,
        price_purchase=0.0,

        # Deprecated: ya no se usa, pero se mantiene por compatibilidad
        discount=0.0,

        margen=0.0,
        iva=0.0,
        iva_value=0.0,

        proveedor=None,
        step_unit=1,

        # Visible para cliente
        is_visible=True,

        # Campos adicionales del sistema
        tipo_pricing=None,
        rate_root=None,
        factor_volumen=None,
        sipsa_id=None,
        last_price_purchase=None,

        status="active",
        created_at=None,
        updated_at=None,
    ):
        self.id = str(id) if id else None

        self.sku = sku
        self.name = name
        self.description = description

        self.category = category
        self.root = root
        self.child = child
        self.unit = unit

        self.price_sale = float(price_sale)
        self.price_purchase = float(price_purchase)
        self.discount = float(discount)  # No se usa para cálculos nuevos

        self.margen = float(margen)
        self.iva = float(iva)
        self.iva_value = float(iva_value)

        self.proveedor = proveedor
        self.step_unit = float(step_unit)

        self.is_visible = bool(is_visible)

        self.tipo_pricing = tipo_pricing
        self.rate_root = rate_root
        self.factor_volumen = factor_volumen
        self.sipsa_id = sipsa_id
        self.last_price_purchase = last_price_purchase

        self.status = status
        self.created_at = created_at or datetime.utcnow()
        self.updated_at = updated_at or datetime.utcnow()


    def to_dict(self):
        return {
            "sku": self.sku,
            "name": self.name,
            "description": self.description,

            "category": self.category,
            "root": self.root,
            "child": self.child,
            "unit": self.unit,

            "price_sale": self.price_sale,
            "price_purchase": self.price_purchase,

            "discount": self.discount,  # deprecated

            "margen": self.margen,
            "iva": self.iva,
            "iva_value": self.iva_value,

            "proveedor": self.proveedor,
            "step_unit": self.step_unit,

            "is_visible": self.is_visible,

            "tipo_pricing": self.tipo_pricing,
            "rate_root": self.rate_root,
            "factor_volumen": self.factor_volumen,
            "sipsa_id": self.sipsa_id,
            "last_price_purchase": self.last_price_purchase,

            "status": self.status,
            "created_at": self.created_at,
            "updated_at": self.updated_at
        }


    def save(self):
        data = self.to_dict()
        result = products_collection.insert_one(data)
        self.id = str(result.inserted_id)
        return self.id


    def update(self):
        if not self.id:
            raise ValueError("Producto sin ID no se puede actualizar")

        self.updated_at = datetime.utcnow()

        products_collection.update_one(
            {"_id": ObjectId(self.id)},
            {"$set": self.to_dict()}
        )


    def delete(self):
        if not self.id:
            return
        products_collection.delete_one({"_id": ObjectId(self.id)})


    @staticmethod
    def find_by_sku(sku):
        doc = products_collection.find_one({"sku": sku})
        if not doc:
            return None
        doc["id"] = str(doc["_id"])
        return doc

    @staticmethod
    def find_by_category(category):
        return list(products_collection.find({"category": category}))

    @staticmethod
    def get_all():
        items = list(products_collection.find())
        for i in items:
            i["id"] = str(i["_id"])
        return items

File: frescapp_web/lib/models/product.dart
class Product {
  final String? id;
  final String? name;
  final String? root;
  final String? child;
  final String? unit;
  final String? category;
  final String? sku;

  final double? priceSale;
  final double? pricePurchase;
  final double? discount;
  final double? margen;
  final bool? iva;
  final double? ivaValue;

  final String? description;
  final String? image;
  final String? status;
  final String? proveedor;

  final double? stepUnit;
  final double? rateRoot;

  /// CANTIDAD MODIFICABLE
  double? quantity;

  // Campos para descuentos
  final double? finalPrice;
  final bool hasDiscount;
  final String? discountType;
  final double? discountValue;
  final double? savingsPct;

  Product({
    this.id,
    this.name,
    this.root,
    this.child,
    this.unit,
    this.category,
    this.sku,
    this.priceSale,
    this.pricePurchase,
    this.discount,
    this.margen,
    this.iva,
    this.ivaValue,
    this.description,
    this.image,
    this.status,
    this.proveedor,
    this.stepUnit,
    this.rateRoot,
    this.quantity,
    this.finalPrice,
    this.hasDiscount = false,
    this.discountType,
    this.discountValue,
    this.savingsPct,
  });

  /// Conversión robusta a double
  static double _toDouble(dynamic v) {
    if (v == null) return 0.0;
    if (v is double) return v;
    if (v is int) return v.toDouble();
    if (v is num) return v.toDouble();
    if (v is String) return double.tryParse(v) ?? 0.0;
    return 0.0;
  }

  /// Conversión segura de IVA
  static bool? _toBool(dynamic v) {
    if (v == null) return null;
    if (v is bool) return v;
    if (v is int) return v == 1;
    if (v is String) return v.toLowerCase() == "true";
    return null;
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json["id"] as String?,
      name: json["name"],
      root: json["root"],
      child: json["child"],
      unit: json["unit"],
      category: json["category"],
      sku: json["sku"],

      priceSale: _toDouble(json["price_sale"]),
      pricePurchase: _toDouble(json["price_purchase"]),
      discount: _toDouble(json["discount"]),
      margen: _toDouble(json["margen"]),
      iva: _toBool(json["iva"]),
      ivaValue: _toDouble(json["iva_value"]),

      description: json["description"],
      image: json["image"],
      status: json["status"],
      proveedor: json["proveedor"],

      stepUnit: _toDouble(json["step_unit"]),
      rateRoot: _toDouble(json["rate_root"]),
      quantity: _toDouble(json["quantity"]),

      // Campos descuento
      finalPrice: _toDouble(json["final_price"] ?? json["price_sale"]),
      hasDiscount: json["has_discount"] ?? false,
      discountType: json["discount_type"],
      discountValue: _toDouble(json["discount_value"]),
      savingsPct: _toDouble(json["savings_pct"]),
    );
  }

  /// Necesario para serializar orders
  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "name": name,
      "root": root,
      "child": child,
      "unit": unit,
      "category": category,
      "sku": sku,

      "price_sale": priceSale,
      "price_purchase": pricePurchase,
      "discount": discount,
      "margen": margen,
      "iva": iva,
      "iva_value": ivaValue,

      "description": description,
      "image": image,
      "status": status,
      "proveedor": proveedor,

      "step_unit": stepUnit,
      "rate_root": rateRoot,

      "quantity": quantity,

      // descuento
      "final_price": finalPrice,
      "has_discount": hasDiscount,
      "discount_type": discountType,
      "discount_value": discountValue,
      "savings_pct": savingsPct,
    };
  }
}
